### ISSUE TITLE
Not able to set default time zone for date picker

### ISSUE THREAD

**Issue by HariSoni87** [Hacker0x01/react-datepicker#1787]
All the dates which we receive from date picker are in local time zone. Is there a way to define the time zone. 
In our application use case we want to get the date in server time zone in place user's time zone. 
React version - 15.6.2
React-datepicker version -  1.6.0

**Comment by neerajtomar98** [comment]
Facing the same issue.

**Comment by stephent** [comment]
As a workaround to this, I'm cheating the selected date by offsetting the actual date/time of my application's state by the utc offsets of the user's default time zone and the time zone my application is working in. Using [moment-timezone.js](https://momentjs.com/timezone/):

```
const mo1 = moment(dateTime); // in browser default time zone, e.g. America/Denver
const mo2 = moment(dateTime).tz(myTimezoneId); // in current state time zone e.g. Africa/Bamako

const modifiedDate = moment(dateTime);
modifiedDate.subtract(mo1.utcOffset() - mo2.utcOffset(), 'minutes');

return modifiedDate.valueOf();
```

Then, in react date picker:

```
<DatePicker selected={modifiedDate} />
```

It would be nice to be able to set a time zone on react-datepicker directly so as to avoid having to do this.

**Comment by JohnStarich** [comment]
I'm doing something similar to @stephent, and wrapped it up into a [component](https://github.com/JohnStarich/sage/blob/b422b0ccef05fffe29143f335017ad2fc24f76fc/web/src/UTCDatePicker.js) to wrap `DatePicker`. So far it's working pretty well, but I'd love for this to be built-in

Happy to contribute this logic if we get some buy-in on the idea.

**Comment by MCBama** [comment]
This is definitely still an issue. Any chance we'll get some word on this?

**Comment by neerajtomar98** [comment]
> This is definitely still an issue. Any chance we'll get some word on this?

Hi please check https://github.com/Hacker0x01/react-datepicker/issues/1388#issuecomment-582590792

**Comment by JohnStarich** [comment]
This seems like a fairly common problem, judging from the top comment reactions.

@martijnrusschen Is this something you want in react-datepicker? If so, I'd be happy to contribute something like a `utc` option to store dates in UTC.

**Comment by danielo515** [comment]
Yep, we really need this too.

**Comment by dzmitry-kankalovich** [comment]
Really need this feature, sometimes I have cases when I need a datepicker to deal with datetime of the specific fixed TZ as opposed to browsers local TZ. 

For example, I live in LA, so it's PST TZ, but I want to select datetime of the schduled job to run in let's say Moscow local TZ.

By default the datetime would be selected for my local PST zone, so my workaround is to offset the selected value to a target TZ in onClick event: 

```javascript
  getTzOffsetMs() {
    const targetTz = +3; // UTC+3
    return (new Date().getTimezoneOffset() + (targetTz * 60)) * 60 * 1000;
  }

  toTargetTime(localTime) {
    // substraction of offset gives results as if I was selecting date having targetTZ as a local TZ in my browser
    return localTime.getTime() - this.getTzOffsetMs() 
  }

  toLocalTime(targetTime) {
    return targetTime.getTime() + this.getTzOffsetMs()
  }
  
  onClickDatepicker(date) {
    const targetTs = this.toTargetTime(date)

    this.setState({
      scheduled_at: targetTs,
      scheduled_at_dp: date.getTime()
    })
  }
```
This though works reliably only if target TZ does not have Daylight Saving Time (DST).

**Comment by swkre** [comment]
Yes @martijnrusschen, it would be great if you can implement this as a built in property! @JohnStarich: Your UTC option would be perfect. We have an app where we would like the client's interaction to be in UTC.

**Comment by stephent** [comment]
@swkre @JohnStarich I think generalizing it beyond just UTC would be preferable. I have a use case where the time zone of the control should match the time zone of a location on the map (which could be anywhere). Being able to set a Timezone ID or UTC offset for the control would be ideal.

**Comment by talovicnedim** [comment]
+1 for UTC.

**Comment by renderlife** [comment]
Нужно TZ

**Comment by JohnStarich** [comment]
@stephent I agree, more flexibility would be good. It seems a `utcOffset` prop used to be available, but may have been removed when the [underlying library changed (#1647)](https://github.com/Hacker0x01/react-datepicker/issues/1647).

For reference, my [wrapper component](https://github.com/JohnStarich/sage/blob/b422b0ccef05fffe29143f335017ad2fc24f76fc/web/src/UTCDatePicker.js) for UTC is fairly simple, so I'll share below. Might be a good starting point for discussion.

If any contributors or maintainers could chime in, _**that would be much appreciated!**_ 🙂

```jsx
import React from 'react';
import DatePicker from 'react-datepicker';


function convertUTCToLocalDate(date) {
  if (!date) {
    return date
  }
  date = new Date(date)
  date = new Date(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate())
  return date
}

function convertLocalToUTCDate(date) {
  if (!date) {
    return date
  }
  date = new Date(date)
  date = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()))
  return date
}

export default function UTCDatePicker({ startDate, endDate, selected, onChange, ...props }) {
  return (
    <DatePicker
      startDate={convertUTCToLocalDate(startDate)}
      endDate={convertUTCToLocalDate(endDate)}
      selected={convertUTCToLocalDate(selected)}
      onChange={date => onChange(convertLocalToUTCDate(date))}
      {...props}
    />
  )
}
```

Related: #1592, #1671, #2043, and possibly #1018.

**Comment by tlehtimaki** [comment]
@JohnStarich how would you go about with setting the possible offset? I guess that would needed to be done internally with react-datepicker?

**Comment by JohnStarich** [comment]
@tlehtimaki Yeah, ideally this would all be done inside react-datepicker somewhere. I'm imagining a return of the `utcOffset` parameter in some fashion could be a good solution. JS has a means of messing with Date objects and time offsets, so should be doable at that level.

Mainly waiting on buy-in from the maintainers before I swoop in with a PR. 🙂

**Comment by tlehtimaki** [comment]
@JohnStarich Thanks for getting back to me! 👍  I was considering if you would be interested to share the idea in a form of fork or the PR as I have many colleagues interested in this too. We are not planning to switch to a fork, but I could review, test and propose changes. So what do you think? I think the former utcOffset only worked if you did not set the date to begin with. I think it would be useful if we could provide a general utcOffset so that:

- date given is first perceived as local
- then it's converted to UTC+0000
- then the offset is applied
- offset is then used to set the context for the date picker
- after selection the time would also show with offset. Not sure how this would work. 🤔

I figure that the feature needs some documentation and I would scope it only to allow setting the offset manually, no full fledged time zone solution.

**Comment by JohnStarich** [comment]
I haven't really dug into the internals of react-datepicker to be honest, though if we think it will make a difference then I can contribute a bit in my spare time 🙂 

That's a great start 👍 As a set of general goals, does this sound reasonable? (Very similar to your list, above.)

* A new `utcOffset` property
* Defaults to local time zone offset. Set to `0` for UTC, `1` for UTC+1h, etc.
* Inputs and outputs are in the same desired offset
* Internal logic can still assume we're using the local time zone (simplifies changes for a PR)

If those are reasonable, I think it might look like this in practice:
* If `utcOffset` is set, input dates (e.g. `startDate`, `endDate`) have their time offsets replaced with the local time zone before storing internally. We could either replace the offset or apply the difference, just need to assess the user experience.
    - Example: If `utcOffset=2` and I set `startDate=2021/02/24 13:45 UTC+02:00` for 1:45pm UTC+2, then it's internally transformed into `startDate=2021/02/24 13:45 UTC-05:00` or 1:45pm local time.
    - I think this will be easiest to use, since it will not require any special conversions from the internal date in end users' code. Input dates are simply assumed to be in the desired offset-time-zone.
* Since the inputs are converted internally, the UI elements will adjust the time based on the UTC offset.
* Output dates (e.g. `onChange` call args) are given an inverse transform from the internal local time zone into the `UTC + utcOffset` form.

Time zone conversions are difficult for me, so I might've gotten these conversions wrong 😅 

Maybe we can try to define the behavior with examples, since those are a little more concrete? In my case, I always want to use UTC on the server side for when a "day" starts at 0m0s. As long as my local dates *appear* to be the same as my server dates, then all is well with the world.

Here's a simplified and converted example from my project, where I'm selecting a month in the UI and need to send start/end dates to the server in UTC.
```jsx
// Use today's local date, but replace time zone with UTC.
function firstOfMonth(date) {
  return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), 1))
}

function lastOfMonth(date) {
  return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth() + 1, 0))
}

const [start, setStart] = React.useState(firstOfMonth(new Date()))
const [end, setEnd] = React.useState(lastOfMonth(new Date()))

React.useEffect(() => {
  API.get('/v1/getBudgets', { params: { start, end } })
    .then(res => res.data.Budgets),
    .then(budgets => /* processing... */ )
}, [start, end])

return (
  <DatePicker
    utcOffset={0} // use UTC+0 for the UI elements
    selected={start} // inputs are in UTC + utcOffset
    onChange={v => { // outputs the selected time in UTC + utcOffset
      setStart(firstOfMonth(v)) // select the whole month using UTC dates
      setEnd(lastOfMonth(v))
    }}
    showMonthYearPicker
  />
)
```
Since it relies on server dates, we just ensure our inputs are always in our desired time zone and the UI will match that time zone with `utcOffset=`.

**Comment by JohnStarich** [comment]
@tlehtimaki, @stephent, @dzmitry-kankalovich would love to hear your thoughts on the above ^

Essentially the idea is to simply select and display dates/times in the chosen UTC offset, so all inputs and outputs do not need to be modified from their server-based sources.

**Comment by tlehtimaki** [comment]
Good thoughts there @JohnStarich, at least the foundations are somewhat clear. JS Date can be initiliazed from UTC, but challenge is that when using it its always in the local time. So with time zone this causes the challenge that we would need to fake the today in order to display it in the desired offset. 

So when my local time would be UTC+3 and I select date at time 01:00 today it would still be yesterday in UTC0 if our offset would be 0, so the challenge would to show the UI according to the offset, not according to the local time. This can be done (I believe) in the way described that we offset all the times with the given number of offset hour/minutes, but somehow ignore the local time zone when displaying times.

What would the problems be we face here? moment solved the problem by handling time in its own objects. Maybe we could do something similar internal and then return the native Date object then as selected value, which user again needs to now that its displayed in local time. 🤔

**Comment by tomhands1** [comment]
Same issue. I would like to always be using UTC if possible

**Comment by JohnStarich** [comment]
@tlehtimaki Makes sense to me 👍 I'm not familiar with the current logic, just anecdotally seems like it's using native Date objects for everything. If so, yep that could be problematic when supporting different output time zones.

If it were me, I think I'd try to store the date in unix time (seconds since Jan 1st 1970 UTC) internally, then convert to Date's when displaying or sending back to the user. That _should_ make it a bit easier to convert between everything.

**Comment by mordgard** [comment]
Would be great if it will be added

**Comment by 55enokky** [comment]
I am looking forward to this fix.

**Comment by martijnrusschen** [comment]
Feel free to open a pull request to make an attempt to fix this.

**Comment by ItzaMi** [comment]
Is there a straight answer to this?
I'm trying to move away from `momentJS` but it seems that any `time` library or even the default JavaScript `Date` has the same issue: everything that is `UTC` is delivered as a string and `ReactDatePicker` still needs a `Date object` so that's no bueno.

I need to get my `Date object` into the `selected` prop as `UTC` but no matter how many conversions I do, I always end up having to use `toDate()` at the end (for the object) which brings me back to `locale time`

**Comment by tmaximini** [comment]
We're using [date-fns-tz](https://www.npmjs.com/package/date-fns-tz) for this, which comes with two helper functions `zonedTimeToUtc ` and `utcToZonedTime` which makes the transformation pretty easy. It returns a date object.

**Comment by DanielPortillo28** [comment]
Hey team

    We also have been facing several issues on this matter, we try to solved but everything turns to be huge work around, implementing a solution out of the component it is not good for production, we do really need this feature.
    
    Thank you

**Comment by curious-33** [comment]
I'm facing the same problem here. When i worked with momentjs, the component would crash with error: Invalid time format. I don't know how to solve this issue

**Comment by tlehtimaki** [comment]
I'm eagerly waiting that we get help for this as a form of a standard. It might be helpful to read the links to understand the problems better so you can circumvent the problems with the time-being.  https://tc39.es/proposal-temporal/docs/index.html https://maggiepint.com/2017/04/09/fixing-javascript-date-getting-started/ Unfortunately we still struggle to have a more permanent solution to this. It is also good to emphasize, that this issue is fundamentally a problem with ECMAScript.

**Comment by VictorJozwicki** [comment]
Wish JS had the Z at the end or you can disable the time in react-datepicker to send only the YYYY-MM-DD ... my backend is setting the date 1 day earlier than it should. I have to send a string without the GMT because of it

**Comment by cwalters162** [comment]
In my current project this has become a recent issue. We want everything to be UTC no matter the client's timezone and looking the the docs folder there was a prop called utcOffset. Unfortunately, it appears to no longer exist in the actual code base. I hope someone has the time to extend date picker with that offset functionality again. It would be very helpful in any project requiring none local client time. 

Ref: https://github.com/Hacker0x01/react-datepicker/blob/master/docs/datepicker.md
| prop | value |
| -- |-- |
| utcOffset | union(number\|string) |

**Comment by nik32** [comment]
This really is needed!!! Need different timezones for a scheduling task. It is becoming a common, important (and somewhat complicated to implement) use case, judging by the above discussions!!!

**Comment by madhavteli** [comment]
We are also facing this issue in several places, need support for different timezones

**Comment by GodAlmighty990** [comment]
Im facing an issue with this aswell

**Comment by GodAlmighty990** [comment]
I havent left my house for 3 days trying to figure out how to fix an issue im having this with datepicker. I am almost done building my project and this is the last issue Im having.


My datePicker is returning a date like this, Im in Colorado.

"Thu Oct 27 2022 02:00:00 GMT-0600 (Mountain Daylight Time)"

When i convert this to UTC i get this 

"2022-10-27T08:00:00Z"

I need to be able to set my timezone to America/New_York, so my datepicker can give me an object like this,

"Thu Oct 27 2022 02:00:00 GMT-0400 (New_York or whatever its actually called)"

This way when I convert to utc I can get this output below

"2022-10-27T06:00:00Z"

Can anyone help me with this?

**Comment by nik32** [comment]
@GodAlmighty990 If its possible to change the component, then I would suggest you to do that. I was having a very hard time integrating timezones with DatePicker. But [mui-x date-tim-picker](https://mui.com/x/react-date-pickers/getting-started/) + [luxon](https://moment.github.io/luxon/#/) really saved the day!! They really made working with timezones a breeze.

**Comment by GodAlmighty990** [comment]
@nik32 
I will look into this component. I am not a huge fan of the clock view when picking times, is it possible to just down a dropdown list of times with this component? thats what I really like about react-datePicker. ex this view
![Screen Shot 2022-10-27 at 1 14 39 PM](https://user-images.githubusercontent.com/74481057/198378634-8fde42db-22cb-4fc7-8165-3e459df7b99d.png)


rather than  a clock with hands

**Comment by nik32** [comment]
@GodAlmighty990 I don't think so... yet I haven't explored this much. The thing was we had a requirement to allow him to choose the timezone for the time. So what really helped was Luxon. It really made it a breeze to work with timezones [coz it give its own DateTime object with timezones supported by the object itself]. And because this mui-x date-time-picker supported luxon [i.e. it could take luxon's DateTime object and display it], it became my go-to solution for timezones!!

**Comment by GodAlmighty990** [comment]
Here is my solution using moment-timezone

my default timezone is "America/Denver"

When I select 2am on October 29th the datepicker returns me a date like this

Sat Oct 29 2022 02:00:00 GMT-0600 (Mountain Daylight Time)

when I use momentt.utc(Sat Oct 29 2022 02:00:00 GMT-0600 (Mountain Daylight Time)).format()

I get 2022-10-29T08:00:00Z 

After I run the function below, If my timezone is "America/New_York" I can console.log the getUTCTime variable and it will return 2022-10-29T06:00:00Z which is the real UTC time for new york at 2am

HOpe this helps someone, If your confused message me I will respond within one day

timezone here is a state, like "America/Chicago" or "America/New_York" make it whatever you want it to be.
```
import momentt from 'moment-timezone';

<DatePicker
 onChange={(date) => { 
                                       const applyOffset = date.setTime(date.getTime() - date.getTimezoneOffset() * 60_000);
                                       const actualTime = new Date(applyOffset).toISOString().replace("Z", "")
                                       const toTz = momentt.tz(actualTime, timezone).format()
                                       const getUTCTime = momentt.utc(toTz).format()

return setUpdatedDateWithTimezone(getUTCTime)
}}
selected={new Date(updatedDateWithTimezone)}
showTimeSelect
showTimeSelectOnly
timeIntervals={60}
timeCaption="Time"
/>

```

**Comment by vivekDebugs** [comment]
@GodAlmighty990, your workaround works well. But how to put `minDate` to time-right-now-in-a-specific-time-zone? The Datepicker only accepts native `Date` object. Even after these fancy conversions, when we do `new Date(getUTCTime)`, it brings the date back to the local time.

**Comment by GodAlmighty990** [comment]
Do const newISO = new Date(getUTCTIme).toISOString()

Then do new Date(newISO) and you should have your date object in the utc time

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by illegalnumbers** [comment]
Is there any progress on this?

**Comment by tannermann** [comment]
Official support for timezone is still something I'd love to see in the docs, even if it is a hack on how to do it.

**Comment by letitechera** [comment]
Having the same issue here! Was this fixed/added?

**Comment by hobadams** [comment]
I'm having the same issue.

**Comment by Mielie** [comment]
I'm working on a project that uses moment-timezone. I've set my Node.js environment timezone to UTC (process.env.TZ="UTC"), I keep my date/times in a moment object  in the desired timezone (`moment().tz(currentZone)`) that is then held in a useState and I convert this to a Date object for react-date-picker by using a string in the constructor that omits the timezone (so 10am Hong Kong time becomes 10am UTC). Then in the onChange function I can push the date/time back into state and force it into my selected timezone:

```
const dateChanged = (date) => {
    const UTCDateTime = moment(date).tz("UTC").format("YYYY-MM-DD[T]HH:mm:ss");
    setSelectedDate(moment.tz(UTCDateTime, currentZone));
};
```

As my environment is set to UTC this should work irrespective of daylight saving times.

To pass the moment object into the date picker:

`selected={new Date(selectedDate.format("YYYY-MM-DD[T]HH:mm:ss"))}`

**Comment by baileygranam** [comment]
I've found a workout by adjusting the display of selected options, as well as the minimum and maximum times, to the user's local time zone. Then, to manage the interaction with a specific target time zone (in this case, Eastern Standard Time or EST), you calculate the difference in hours between the user's local time zone and EST. Based on this calculated difference, you either add or subtract the hours to align the times with the desired time zone.

This effectively accounts for the time zone difference, ensuring that the time inputs and displays are relevant both to the user's local context and the target time zone context. This workaround allows users to work in their local time zone while ensuring the underlying data is synchronized with the target time zone. 

```typescript
import React from 'react';
import ReactDatePicker, { ReactDatePickerProps } from 'react-datepicker';
import moment from 'moment-timezone';

export type TimeInputProps = {
  onChange?: (value: Date | null) => void;
} & ReactDatePickerProps;

export const TimeInput: React.FC<TimeInputProps> = ({
  selected,
  minTime,
  maxTime,
  onChange,
  ...rest
}) => {
  const userTimezone = moment.tz.guess();

  const userTimezoneOffsetMinutes = moment.tz
    .zone(userTimezone)
    ?.utcOffset(moment().valueOf());

  const estTimezoneOffsetMinutes = moment.tz
    .zone('America/New_York')
    ?.utcOffset(moment().valueOf());

  const hoursDifferenceToEST =
    ((estTimezoneOffsetMinutes ?? 0) - (userTimezoneOffsetMinutes ?? 0)) / 60;

  const adjustDateToEST = (date: Date | null) => {
    if (date) {
      onChange?.(moment(date).add(hoursDifferenceToEST, 'hours').toDate());
    }

    return null;
  };

  const convertDateToLocalTimezone = (date?: Date | null) => {
    return date
      ? moment(date).subtract(hoursDifferenceToEST, 'hours').toDate()
      : undefined;
  };

  return (
    <ReactDatePicker
      timeIntervals={15}
      showTimeSelect
      showTimeSelectOnly
      dateFormat="h:mm aa"
      selected={convertDateToLocalTimezone(selected)}
      minTime={convertDateToLocalTimezone(minTime)}
      maxTime={convertDateToLocalTimezone(maxTime)}
      onChange={adjustDateToEST}
      {...rest}
    />
  );
};
```

**Comment by jnorvell** [comment]
I need the DatePicker (and time picker) to display and input selection in an arbitrary timezone that is not the user's local machine timezone.

Providing it as a prop and letting the library do all other conversions for display et c. seems required.

```
<DatePicker
   timeZone="US/Pacific"
/>

```

**Comment by zacheryph** [comment]
This is truly amazing, and in large part no fault of this library. I find date handling in Javascript astounding... The fact that you cannot represent a _DATE_ in Javascript, or use a simple _DATE_ Picker without reading a 5 year old issue thread is a remarkable feat in this day and age. Now some may say "dates without time don't mean alot" but I disagree. "New Years Day" does not have a timezone. It is 1/1. The fact that it "becomes" new years day at "midnight" is besides the point. Wherever you are, it is new years day. All day. Not being able to reliably represent something as simple as this is just... mind boggling. More importantly, when using a _DATE_ picker, I should be able to reliably handle & format that _DATE_ without having to worry about it magically changing. (I know.. JS's _Date_ object is to blame...)

My _"question"_ probably doesn't even belong here but ...

`<DatePicker selected={date} />`

Now when `date="01/01/2025"` The date picker luckily says the date of new years... But if `date="2025-01-01"` I am greeted by the wonderful date of New Years Eve... This I assume is cause the library is simply using `new Date()` to parse the given string since it is not already a Date object.

Given this assumption, I cannot really _blame_ the library for simply using the underlying language, but _some_ of the blame can still be placed I'd say. I cannot  give a _DATE_ picker a date, and reliably have it display that date, without having to worry about timezones? 🤯 

My real question is.... How in the hell do I pick a _DATE_ and reliably represent it on the picker, in code, and format it for form data as the same value?!?!?!

**Comment by tannermann** [comment]
@zacheryph your incredulity at the longstanding nature of this annoyance is warranted. It is worth holding out hope, however. JS temporals are on the horizon and may make these troubles fade out of existence over _time_. [You can read more about temporals here](https://tc39.es/proposal-temporal/docs/).

**Comment by stephent** [comment]
@tannermann @zacheryph although, having said that, `date-fns` on which this component depends added time zone support in v4 a couple of months ago, so there's really no fundamental barrier to implementing a decent solution now. 

https://blog.date-fns.org/v40-with-time-zone-support/

**Comment by stephent** [comment]
It seems react-datepicker is using `date-fns@3.6.0`, so no time zone support for now. I've switched to `react-day-picker` instead which is using `date-fns@4.1` already.

**Comment by minhlucvan** [comment]
I am facing the same issue, please consider implementing this as soon as possible! 🙏

**Comment by afilp** [comment]
@mindplay-dk  Interersting approach, how will the min/max options be supported though? Or other similar props that need to know about timeZone? Thanks!

**Comment by mindplay-dk** [comment]
Good question.

I suppose you could derive those values in the same way.

As long as you initialize them correctly - that is, with a `T00:00:00.000Z` time setting.

The datepicker uses `date-fns` for all of it's date arithmetic and comparisons, should that should "just work".

The biggest stumbling block for me thus far has been the cases where you want to initialize a date-only picker to today's date - this is really quite tricky, because developers will intuitively initialize these as just `new Date()`, right? But in some time zones this will be the wrong date *in the UTC time zone*.

As a workaround, I came up with this horror in our date-picker wrapper component:

```ts
/**
 * This function uses a heuristic to see if the given Date was most
 * likely generated using `new Date()`, and if so, takes the date
 * in the user's time zone and creates that date in UTC at 00:00:00.
 *
 * Because we store date-only values as UTC 00:00:00 timestamps, we
 * expect the input to this function to be either e.g.:
 *
 * - `new Date('2020-01-01T00:00:00.000Z')`, which the function
 *   would leave as-is and merely wrap it in a TZDate, so that the
 *   date-picker can display is correctly.
 *
 * - `new Date()`, which the function would recognize is almost
 *   definitely a date that was generated on the client, and must
 *   be converted to a UTC 00:00:00 timestamp representing the
 *   date in the user's time zone.
 *
 * For example, if given a Date value `2025-04-28T12:33:33.333Z`, if
 * the date was generated in Auckland, the function produces a new
 * `Date` instance with the ISO value `2025-04-29T00:00:00.000Z`,
 * since the 28th UTC was already the 29th in Auckland at that time.
 *
 * If this makes your brain hurt, I am sorry. The alternative was
 * to introduce a `UTCDateNow()` function and use that to generate
 * a UTC 00:00:00 date at any call site where the default value of
 * a date-only input is currently (incorrectly) set to `new Date()`,
 * which would have been far more error prone.
 *
 * This way, we get something that works for both date-only and
 * date/time pickers, and where an intuitive (albeit incorrect)
 * default value of `new Date()` "just works".
 */
function toUTCDate(date: Date): Date {
  const isUTCMidnight =
    date.getUTCHours() === 0 &&
    date.getUTCMinutes() === 0 &&
    date.getUTCSeconds() === 0 &&
    date.getUTCMilliseconds() === 0

  return isUTCMidnight
    ? new TZDate(date, 'UTC')
    : new TZDate(
        Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()),
        'UTC',
      )
}
```

Don't say it. I know. I'm not proud of this.

But the odds of someone actually generating a `new Date()` at precisely 00:00:00.000 is highly unlikely.

And the alternative was the potential of introducing possibly hundreds of subtle bugs throughout the code base.

Alternatively, you could of course write a `todayUTC` function that generates the user's date in the UTC timezone - but you'd have to teach your entire team when and how and why they need to use this, which... is that even obvious from everything I just explained?

Sometimes there are no good or easy answers. 😌

**Comment by tyteen4a03** [comment]
Still having the same issue.

**Comment by martijnrusschen** [comment]
I've created a PR to add support for adding a timezone to the datepicker. I'm curious to hear your thoughts on this implementation before we go down this path: https://github.com/Hacker0x01/react-datepicker/pull/6068. Please check it out and let me know if this would be a helpful improvement.
